### ISSUE TITLE
Reduce null's aggressiveness in insert mode

### ISSUE THREAD

**Issue by mystilleef** [jose-elias-alvarez/null-ls.nvim#177]
### Issues

- [X] I have checked [existing issues](https://github.com/jose-elias-alvarez/null-ls.nvim/issues) and there are no existing ones with the same request.

### Feature description

Is there a way to turn off `null` temporarily when in insert mode?

There's an issue with the `markdown` linter that's requiring a confirmation on every `CursorHoldI`.

The screenshot of the error is below.

For performance and distraction reasons, some users prefer diagnostic and linting checks to trigger in `normal` mode instead of `insert` mode.

Would you consider providing an option for this behavior?

## Screenshot

![nullls](https://user-images.githubusercontent.com/273399/132774014-83a4ce20-c4bc-42ae-8304-3faf3716acd6.png)

## Neovim version

```
NVIM v0.6.0-dev+234-g2fe95cb1a
Build type: Release
LuaJIT 2.1.0-beta3
Compiled by lateef@bigdaddy
Features: +acl +iconv +tui
system vimrc file: "$VIM/sysinit.vim"
fall-back for $VIM: "/usr/local/share/nvim"
```

### Help

No

**Comment by jose-elias-alvarez** [comment]
The specific issue you’re seeing is most likely related to #172 and #174. Once that’s fixed, you shouldn’t see errors like that unless the linter is misconfigured, regardless of the mode or the file’s content.

Separately, I think updating diagnostics only in normal mode would be a great enhancement, and I think it lines up with user expectations more closely (confusingly, Neovim’s `update_in_insert` flag only affects the display of LSP diagnostics and not when the editor sends notifications). You can see an attempt I made to implement this in #136, but my conclusion was that the approach I took there caused more issues. Once I have more time, I’m considering opening an issue upstream to gauge interest in implementing this in Neovim core.

**Comment by mystilleef** [comment]
I can confirm that I use multiple linters for markdown files, `vale` and `markdownlint`.

So the issue may be related to how multiple `null` linters attached to a single buffer are managed.

Either way, this issue is a recent development. I never had this problem in the past.

In general, other linters, at least the native ones, don't trigger in insert mode. I don't remember if I set it up this way, or if this is the default behavior. 

I deliberately set my `updatetime` to 1000 and `debounce` timeout to 1500 to reduce the general aggressiveness of Neovim and some plugins. It's been great for my sanity. :-D

The recommended best practice for plugins is to do little to no processing in insert mode. I don't know if it already does, but it'll be great if `null` could adopt this stance by default.

**Comment by jose-elias-alvarez** [comment]
I also suspect it’s a regression since I think we would have had reports of this in the past, so I plan on investigating more. We have a fix available at #179 for now, so could you check if that solves the error popup?

My understanding is that Neovim will send LSP notifications and process responses while in insert mode, regardless of the user’s settings, so the current behavior is a consequence of that, and I believe both efm and diagnostic-languageserver work the same way. I have some other ideas for avoiding updating in insert mode that I plan on trying out when I have time / access to a computer again.

**Comment by jose-elias-alvarez** [comment]
The error should be fixed by #198, and I'm using #199 to track progress on the insert mode issue, so I'm going to close this for now.
